scripts/service.py

```diff
@@ -0,0 +1,286 @@
+        try:
+            subprocess.Popen(self.script)
+        except OSError as err:
+            self.logErr("execute script error:" + str(err))
+            
+    def _readRequest(self):
+        """ read all that is written in the pipe, so in case of input storm,
+            just do work once.
+        """
+        if self.fifo != None:
+            try:
+                while True:
+                    self.buffer = os.read(self.fifo, BUFFER_SIZE)
+                    if len(self.buffer) == 0:
+                        # hung up
+                        self.logErr("hung up found while read, exiting")
+                        raise IOError(errno.EPIPE, 'hung up occurred while read pipe', self.pipePath)
+            except OSError as err:
+                if err.errno == errno.EAGAIN or err.errno == errno.EWOULDBLOCK:
+                    return True # assume already read as invoked after poll
+                else:
+                    self.logErr("Error occurred while read pipe:" + str(err))
+                    raise
+        else:
+            return False
+            
+    def _waitForInput(self):
+        """ hang on the poll to wait for incoming requests.
+            As we hold one writer, there should be no EPOLLHUP when others closed the pipe.
+        """
+        try:
+            events = self.epollObj.poll(-1)
+        except KeyboardInterrupt:
+            return False
+        except IOError as err:
+            if err.errno == errno.EINTR:
+                return False
+            else:
+                self.logErr("Error at polling fd:" + str(err))
+                raise
+
+        for fileno, event in events:
+            if fileno == self.fifo and event & select.EPOLLIN:
+                return True
+            elif fileno == self.fifo and event & (select.EPOLLERR | select.EPOLLHUP):
+                # probably some log and raise failure
+                self.logErr("exit due to unexpected events:" + str(event))
+                raise IOError(errno.EPIPE, 'hung up or error occurred while poll', self.pipePath)
+
+        return False
+                
+    def work(self):
+        """ server side daemon logic """
+        self.logInfo("server started")
+        
+        try:
+            self._openRead()
+            while True:
+                if self._waitForInput():
+                    if self._readRequest():
+                        self._executeSub()
+        except Exception as err:
+            self.logErr("Exit at error:" + str(err))
+            self._closeFIFO()
+
+
+class ClientInstance(ServiceSkeleton):
+    """ client side implementation, to invoke the server/daemon to do work """
+    def __init__(self, name):
+        super(ClientInstance, self).__init__(name)
+        
+    def _writeRequest(self):
+        """ write the pipe to wake up the daemon to do the work """
+        if self.fifo == None:
+            self._openWrite()
+            
+        try:
+            os.write(self.fifo, self.name)
```

Comment:
Since we actually care what is written through the pipe, only that a write happens, why write more than one byte? Takes longer to fill the pipe by accident that way.

Reply:
@TimSmithCtx , This actually leaves some space for further more trick check to verify the written is valid. Since the service name will be very short, so the time taken will not be different. The OS will do read/write per what's 'cached' in the buffer of the pipe, so written one bytes and several actually take the same time.

Reply:
You're not taking a lock on the FIFO when you open it for write. If udev fires off two of your workers in close succession called "myeventname", and you read "myevmyeventnameentname" in the server, how will you validate it? :-)

Reply:
It is not kind of byte level partition. For PIPE, each write is done per one pipe buffer, which should be 4K or 64K bytes (also in that case, one byte or several bytes does not have performance difference), so if we just use some small string as service name, it wouldn't get split/mixed. But any way, it is just a potential, and we can leave it there.

Reply:
write() is not atomic, so if you write more than 1 byte at a time from 2 processes to the same file/pipe without locking, at some point they *will* get mixed. It's not important now but would be if we tried to use that.

Reply:
@TimSmithCtx , yes, write() itself is not atomic, but this is write to a pipe. Please refer to the following for details, as what I described above:
From http://man7.org/linux/man-pages/man7/pipe.7.html:
==========================
POSIX.1 says that write(2)s of less than PIPE_BUF bytes must be
       atomic: the output data is written to the pipe as a contiguous
       sequence.  Writes of more than PIPE_BUF bytes may be nonatomic: the
       kernel may interleave the data with data written by other processes.
       POSIX.1 requires PIPE_BUF to be at least 512 bytes.  (On Linux,
       PIPE_BUF is 4096 bytes.) 
==========================
But anyway, we still did not do that, so let's just not wasting more time discussing this.

Reply:
Yes, fair enough.